### PR TITLE
Remove unused hash strategies

### LINKED ISSUE #1498: `createtree` default for `hash_strategy` (RFC6962_SHA256) is `an unknown hasher`
See #901 where this was reported but not resolved.

```
go run github.com/google/trillian/cmd/createtree   --help
...
  -hash_algorithm string
    	Hash algorithm of the new tree (default "SHA256")
  -hash_strategy string
    	Hash strategy (aka preimage protection) of the new tree (default "RFC6962_SHA256")
...
```
But:
```
go run github.com/google/trillian/cmd/createtree \
  --admin_server=localhost:8095 \
  --tree_type=MAP
E0405 12:12:34.638524  220463 admin.go:64] failed to CreateTree(tree:<tree_state:ACTIVE tree_type:MAP hash_strategy:RFC6962_SHA256 hash_algorithm:SHA256 signature_algorithm:ECDSA max_root_duration:<> > key_spec:<ecdsa_params:<> > ): *status.statusError rpc error: code = InvalidArgument desc = failed to create hasher for tree: MapHasher(RFC6962_SHA256) is an unknown hasher
F0405 12:12:34.638878  220463 main.go:176] Failed to create tree: rpc error: code = InvalidArgument desc = failed to create hasher for tree: MapHasher(RFC6962_SHA256) is an unknown hasher
exit status 1
```
The command requires:
```
--hash_strategy=TEST_MAP_HASHER
```

### PR DESCRIPTION
**Pull request by gdbelvin**
The following hashers have been removed because of security vulnerabilities and performance issues:

- `OBJECT_RFC6962_SHA256` - Security vulnerability
- `CONIKS_SHA256` - Less than ideal performance 

- [x] I have updated the [CHANGELOG](CHANGELOG.md).
  - Adjust the draft version number according to [semantic versioning](https://semver.org/) rules.
- [x] I have updated [documentation](docs/) accordingly (including the [feature implementation matrix](docs/Feature_Implementation_Matrix.md)).

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -4,6 +4,11 @@
 
 Not yet released; provisionally v1.4.0 (may change).
 
+### Remove Unsupported Hashers
+
+- `OBJECT_RFC6962_SHA256`
+- `CONIKS_SHA256`
+
```
Comment (Martin2112) [inline CHANGELOG.md]: Probably worth adding the reason from the PR description here.

File: integration/maptest/map.go [hunk 2]
```diff
@@ -172,8 +172,7 @@ func RunMapRevisionZero(ctx context.Context, t *testing.T, tadmin trillian.Trill
 		{
 			desc: "empty map has SMR at rev 0 but not rev 1",
 			hashStrategy: []hashStrategyAndRoot{
```
Comment (pavelkalinnikov) [inline integration/maptest/map.go]: This doesn't have to be a slice anymore. Add a TODO to simplify this test?

Comment (pavelkalinnikov) [reply integration/maptest/map.go]: Same argument applies to all the tests in this PR that are left with just one hash strategy.

File: merkle/hstar2_test.go [hunk 3]
```diff
@@ -35,22 +34,15 @@ const treeID = int64(0)
 
 var (
 	deB64 = testonly.MustDecodeBase64
```
Comment (pavelkalinnikov) [inline merkle/hstar2_test.go]: Is `deB64` used anywhere else now?

File: merkle/hstar2_test.go [hunk 4]
```diff
@@ -35,22 +34,15 @@ const treeID = int64(0)
...
-// rootB64 is the incremental root after adding the corresponding i/v pair, and
-// all i/v pairs which come before it.
-//
-// Values were calculated with the C++/Python sparse Merkle tree code in the
-// github.com/google/certificate-transparency repo.
-// TODO(alcutter): replace with hash-dependent computation. How is this computed?
 var simpleTestVector = []struct {
 	index, value, root []byte
 }{
-	{nil, nil, deB64("xmifEIEqCYCXbZUz2Dh1KCFmFZVn7DUVVxbBQTr1PWo=")}, // Empty tree.
```
Comment (pavelkalinnikov) [inline merkle/hstar2_test.go]: Is there a reason to delete the empty tree test? I guess, it should also have some "empty tree" hash value.

**Comment by gdbelvin** [comment]
Moved the `TEST_MAP_HASHER` change to its own PR

File: CHANGELOG.md [hunk 5]
```diff
@@ -4,6 +4,15 @@
 
 Not yet released; provisionally v1.4.0 (may change).
 
+### Remove Unsupported Hashers
+
+The following hashers have been removed because of security vulnerabilities and
+performance issues:
+
+- `OBJECT_RFC6962_SHA256` - Security vulnerability
```
Comment (mhutchinson) [inline CHANGELOG.md]: A link to some external info on this security vulnerability would be useful, if it exists.

File: CHANGELOG.md [hunk 6]
```diff
@@ -4,6 +4,15 @@
...
+- `CONIKS_SHA256` - Less than ideal performance
```
Comment (mhutchinson) [inline CHANGELOG.md]: Add a recommendation that anyone using this ought to migrate to the 512/256 version?

File: integration/maptest/map.go (hunk 2 above)
Comment (gdbelvin) [reply integration/maptest/map.go]: I'll remove the slice in the next PR which removes TEST_MAP_HASHER

File: merkle/hstar2_test.go (hunk 3 above)
Comment (gdbelvin) [reply merkle/hstar2_test.go]: yes

**Comment by gdbelvin** [comment]
The reason for this PR is mildly related to #1498 
By removing unused hashers we can make the tree utilities work seamlessly. 
Not a high priority.  Should I close this PR and wait for semantic versioning?

**Comment by mhutchinson** [comment]
> Not a high priority. Should I close this PR and wait for semantic versioning?

Thanks for the context. My instinct is to update the docs to mark these as deprecated but not make this code-breaking change right now. But I'm more averse than most to making breaking changes.

**Pull request closed without merging**